File: GameLibrary/src/cc/lib/game/IGame.java
package cc.lib.game;

import java.util.List;

/**
 * Interface for a MiniMaxDescision Tree
 * @param <M>
 */
public interface IGame<M extends IMove> {
    /**
     * This will push of the game
     * @param move
     */
    void executeMove(M move);

    /**
     * This will pop state
     * @return
     */
    M undo();

    /**
     * Get all the moves available for this state
     * @return
     */
    List<M> getMoves();

    /**
     * Get the current player turn
     * @return
     */
    int getTurn();

    int getWinnerNum();

    boolean isDraw();
}

File: GameLibrary2/src/cc/lib/game/IGame.kt
package cc.lib.game

/**
 * Interface for a MiniMaxDescision Tree
 * @param <M>
</M> */
interface IGame<M : IMove> {
	/**
	 * This will push of the game
	 * @param move
	 */
	fun executeMove(move: M)

	/**
	 * This will pop state
	 * @return
	 */
	fun undo(): M

	/**
	 * Get all the moves available for this state
	 * @return
	 */
	fun getMoves(): List<M>

	/**
	 * Get the current player turn
	 * @return
	 */
	var turn: Int
	fun getWinnerNum(): Int
	fun isDraw(): Boolean
}